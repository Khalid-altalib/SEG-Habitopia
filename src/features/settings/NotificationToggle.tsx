import { Box, Switch, Text } from "native-base";
import React from "react";

type Props = {};

const NotificationToggle = (props: Props) => {
  return (
    <Box display="flex" justifyContent={"space-between"}>
      <Text>Recieve In-App Notifications </Text>
      <Switch size="md"></Switch>
    </Box>
  );
};

export default NotificationToggle;
