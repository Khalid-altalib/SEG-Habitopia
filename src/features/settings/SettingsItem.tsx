import React from "react";
import { Box, Pressable, Text } from "native-base";

type Props = {
  type: string;
};

function SettingsItem(props: Props) {
  const { type } = props;
  return (
    <Pressable>
      <Text>{type}: placeholder</Text>
    </Pressable>
  );
}

export default SettingsItem;
