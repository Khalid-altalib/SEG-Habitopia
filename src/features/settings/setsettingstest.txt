import {
  Action,
  AnyAction,
  configureStore,
  Dispatch,
  ThunkAction,
} from "@reduxjs/toolkit";
import settingsReducer, { fetchSettings, setSettings } from "./settingsSlice";
import { getUserFromDatabase } from "@app/util";

jest.mock("@app/util", () => ({
  getUserFromDatabase: jest.fn(),
}));

describe("settingsSlice", () => {
  let store: ReturnType<
    typeof configureStore & {
      dispatch: Dispatch<AnyAction>;
    }
  >;
  const mockUserData = {
    email: "test@test.com",
    name: "Test User",
    notifications: true,
    biography: "This is a test biography.",
  };

  beforeEach(() => {
    store = configureStore({
      reducer: {
        settings: settingsReducer,
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("setSettings", () => {
    const mockSettings = {
      name: "Test User",
      notifications: true,
      biography: "This is a test biography.",
      password: "newPassword",
      oldPassword: "oldPassword",
    };

    it("should handle setSettings correctly", async () => {
      const expectedSettings = {
        email: "",
        password: "",
        notifications: mockSettings.notifications,
        name: mockSettings.name,
        biography: mockSettings.biography,
      };

      try {
        await store.dispatch(setSettings(mockSettings));
      } catch (error) {
        console.log(error);
      }

      const actualSettings = store.getState().settings.settings;
      expect(actualSettings).toEqual(expectedSettings);
    });

    it("should handle setSettings with only password update", async () => {
      const mockPasswordUpdate = {
        password: "newPassword",
        oldPassword: "oldPassword",
      };

      await store.dispatch(setSettings(mockPasswordUpdate));

      const actualSettings = store.getState().settings.settings;
      expect(actualSettings.password).toEqual(mockPasswordUpdate.password);
    });

    it("should handle setSettings with errors", async () => {
      const errorMessage = "Error setting user data";
      (getUserFromDatabase as jest.Mock).mockRejectedValueOnce(
        new Error(errorMessage)
      );

      await expect(
        store.dispatch(setSettings(mockSettings))
      ).resolves.toMatchObject({
        error: expect.objectContaining({ message: "Rejected" }),
      });
    });
  });
});

type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
