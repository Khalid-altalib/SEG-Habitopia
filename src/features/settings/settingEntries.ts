type SettingEntries = {
  [key: string]: SettingEntry;
};

export type SettingEntry = {
  name: string;
};

const settingEntries: SettingEntries = {
  name: {
    name: "Name",
  },
  password: {
    name: "Password",
  },
  biography: {
    name: "Bio",
  },
};

export default settingEntries;
