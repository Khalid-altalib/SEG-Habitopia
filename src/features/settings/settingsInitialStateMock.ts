const initialState = {
  settings: {
    email: "example@gmail.com",
    password: "",
    notifications: true,
    name: "John Doe",
    biography: "hello my name is john doe and i have severe autism",
  },
  fetchSettings: {
    loading: false,
    error: "",
  },
  setSettings: {
    loading: false,
    error: "",
  },
};

export default initialState;
