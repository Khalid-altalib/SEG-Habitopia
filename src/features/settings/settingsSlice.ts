import { createSlice, createAsyncThunk, PayloadAction } from "@reduxjs/toolkit";
import { DataStore } from "aws-amplify";
import { Toast } from "react-native-toast-message/lib/src/Toast";
import { Settings } from "../../../types";
import { getUserFromDatabase } from "../../app/util";
import { User } from "../../models";
import { updatePassword } from "../auth/authSlice";

export type SettingsState = {
  settings: Settings;
  fetchSettings: {
    loading: boolean;
    error: string;
  };
  setSettings: {
    loading: boolean;
    error: string;
  };
};

const initialState: SettingsState = {
  settings: {
    email: "",
    password: "",
    notifications: false,
    name: "",
    biography: "",
  },
  fetchSettings: {
    loading: false,
    error: "",
  },
  setSettings: {
    loading: false,
    error: "",
  },
};

export const fetchSettings = createAsyncThunk<
  Settings,
  void,
  { rejectValue: string }
>("settings/fetch", async (_, thunkAPI) => {
  try {
    const userData  = await getUserFromDatabase(thunkAPI);

    const response =  {
      email: userData.email,
      name: userData.name,
      notifications: userData.notifications,
      biography: userData.biography,
    };
   
    return (await response) as Settings;
  } catch (error: any) {
    const message = error.message;
    return thunkAPI.rejectWithValue(message);
  }
});


export const setSettings = createAsyncThunk<
  void,
  object,
  { rejectValue: string }
>("settings/set", async (settings: any, thunkAPI) => {
  try {
    const {name, notifications, biography, password, oldPassword} = settings;
    const user = await getUserFromDatabase(thunkAPI);
    // update the one that is not null
    await DataStore.save(
      User.copyOf(user, (updated) => {
        if (name !== undefined) {
          updated.name = name;
        }
        if (notifications !== undefined) {
          updated.notifications = notifications;
        }
        if (biography !== undefined) {
          updated.biography = biography;
        }
        if (password !== undefined) {
          updatePassword(password, oldPassword);
        }
      }));

    Toast.show({type: "success", text1: "Your new setting has been saved! âœ…"})

    return settings; 
  } catch (error: any) {
    const message = error.message;
    return thunkAPI.rejectWithValue(message);
  }
});

export const settingsSlice = createSlice({
  name: "settings",
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(
      fetchSettings.fulfilled,
      (state: SettingsState, action: PayloadAction<Settings>) => {
        state.settings = action.payload;
        state.fetchSettings.loading = false;
        state.fetchSettings.error = "";
      }
    );
    builder.addCase(fetchSettings.pending, (state: SettingsState) => {
      state.fetchSettings.loading = true;
      state.fetchSettings.error = "";
    });
    builder.addCase(
      fetchSettings.rejected,
      (state: SettingsState, action: PayloadAction<any>) => {
        state.fetchSettings.loading = false;
        state.fetchSettings.error = action.payload.message;
      }
    );
    builder.addCase(
      setSettings.fulfilled,
      (state: SettingsState, action: PayloadAction<any>) => {
        state.settings = { ...state.settings, ...action.payload };
        state.setSettings.loading = false;
      }
    );
    builder.addCase(setSettings.pending, (state: SettingsState) => {
      state.setSettings.loading = true;
    });
    builder.addCase(setSettings.rejected, (state: SettingsState) => {
      state.setSettings.loading = false;
    });
  },
});

export const {} = settingsSlice.actions;

export default settingsSlice.reducer;
