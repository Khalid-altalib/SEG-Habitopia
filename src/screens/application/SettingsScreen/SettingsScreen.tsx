import React from "react";
import SettingsPage from "@features/settings/SettingsPage/SettingsPage";
import { ScrollView } from "native-base";
import Background from "@components/Background";
import PaddedContainer from "@components/PaddedContainer";

const SettingsScreen = () => {
  return (
    <Background>
      <PaddedContainer>
        <ScrollView>
          <SettingsPage />
        </ScrollView>
      </PaddedContainer>
    </Background>
  );
};

export default SettingsScreen;
