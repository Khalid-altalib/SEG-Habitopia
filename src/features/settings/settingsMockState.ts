import { SettingsState } from "./settingsSlice";

/**
 * The mock state for the settings redux state
 */

const settingsMockState: SettingsState = {
  settings: {
    email: "test@test.com",
    password: "password",
    notifications: true,
    name: "Test User",
    biography: "Hello, I'm Test User.",
  },
  fetchSettings: {
    loading: false,
    error: "",
  },
  setSettings: {
    loading: false,
    error: "",
  },
};

export default settingsMockState;
