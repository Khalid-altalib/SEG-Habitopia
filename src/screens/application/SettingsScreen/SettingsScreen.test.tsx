// React
import React from "react";

// React Native Testing Library
import { render, screen } from "@testing-library/react-native";

// Redux
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

// Habitopia
import SettingsScreen from "@screens/application/SettingsScreen";
import TestingWrapper from "@app/testingWrapper";
import { SettingsState } from "@features/settings/settingsSlice";

it("Renders correctly.", () => {
  const mockSettingsState: SettingsState = {
    settings: {
      email: "",
      password: "",
      notifications: false,
      name: "",
      biography: "",
    },
    fetchSettings: {
      loading: false,
      error: "",
    },
    setSettings: {
      loading: false,
      error: "",
    },
  };

  const mockStore = configureStore([thunk])({
    settings: mockSettingsState,
  });

  render(
    <TestingWrapper store={mockStore}>
      <SettingsScreen />
    </TestingWrapper>
  );

  expect(screen.toJSON()).toMatchSnapshot();
});
