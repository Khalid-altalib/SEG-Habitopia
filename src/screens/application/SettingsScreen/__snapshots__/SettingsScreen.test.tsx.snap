// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly. 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      {
        "colors": [
          4293190887,
          4294243573,
        ],
        "endPoint": undefined,
        "locations": undefined,
        "startPoint": undefined,
      }
    }
    style={
      {
        "height": "100%",
      }
    }
  >
    <RCTScrollView
      contentContainerStyle={
        [
          {},
          {
            "dataSet": {},
          },
          {},
        ]
      }
      dataSet={{}}
      style={
        {
          "height": "100%",
        }
      }
    >
      <View>
        <View
          style={
            {
              "height": "100%",
              "padding": 25,
              "width": "100%",
            }
          }
        >
          <View
            dataSet={{}}
            style={{}}
          >
            <View
              dataSet={{}}
              style={
                {
                  "alignItems": "center",
                  "display": "flex",
                  "justifyContent": "center",
                }
              }
            />
          </View>
        </View>
      </View>
    </RCTScrollView>
  </ViewManagerAdapter_ExpoLinearGradient>
</RNCSafeAreaProvider>
`;
